src/amplify-migration/amplifyMigrationScript.ts

```diff
@@ -0,0 +1,388 @@
+  if (matches) {
+    matches.forEach((match: string) => {
+      match = getRawPermalink(match);
+      if (changedPermalinks[match]) {
+        navigationFileChanged = true;
+        navigationYmlContent = navigationYmlContent.replace(
+          match,
+          changedPermalinks[match]
+        );
+      }
+    });
+  }
+  if (navigationFileChanged) {
+    await fs.promises.writeFile(
+      navigationYmlPath,
+      navigationYmlContent,
+      "utf-8"
+    );
+    await simpleGit(repoPath).add(navigationYmlPath);
+  }
+}
+
+/**
+ * Requirements:
+ * convert `/some/path` to `some/path`
+ * convert `some/path/` to `some/path`
+ * above two variants with `permalink: ` prefix
+ * above two variants with `url: ` prefix
+ * @param permalink original permalink
+ * @returns raw permalinks without leading/trailing slash
+ */
+function getRawPermalink(permalink: string) {
+  let trimmedPermalink = permalink.trim();
+  if (trimmedPermalink.startsWith(`permalink: `)) {
+    trimmedPermalink = permalink.trim().slice(11);
+  }
+  if (trimmedPermalink.startsWith(`url: `)) {
+    trimmedPermalink = permalink.trim().slice(5);
+  }
+  if (trimmedPermalink.startsWith(`/`)) {
+    trimmedPermalink = trimmedPermalink.slice(1);
+  }
+  if (trimmedPermalink.endsWith("/")) {
+    trimmedPermalink = trimmedPermalink.slice(0, -1);
+  }
+
+  return trimmedPermalink;
+}
+
+async function changePermalinksInMdFile(
+  filePath: string,
+  repoPath: string,
+  changedPermalinks: { [key: string]: string }
+) {
+  let fileContent = await fs.promises.readFile(filePath, "utf-8");
+  let fileChanged = false;
+
+  ({ fileContent, fileChanged } = changeFileContent(
+    fileContent,
+    changedPermalinks
+  ));
+
+  if (fileChanged) {
+    await fs.promises.writeFile(filePath, fileContent, "utf-8");
+    await simpleGit(repoPath).add(filePath);
+  }
+}
+
+function changeFileContent(
+  fileContent: string,
+  changedPermalinks: { [oldPermalink: string]: string }
+) {
+  let hasFileChanged = false;
+  // two different permalink patterns to take care of
+  // 1. href="original_permalink"
+  // 2. [click here](original_permalink)
+  // 1 is solved using JSDOM, 2 is solved using regex
+
+  const markdownRegex = /\[(.*?)\]\((.*?)\)/g;
+  const dom = new JSDOM(fileContent);
+  dom.window.document.querySelectorAll("a").forEach((a: any) => {
+    // replace permalinks with lowercase and in changedPermalinks
+    let rawPermalink = getRawPermalink(a.href);
+
+    if (changedPermalinks[rawPermalink]) {
+      a.href = a.href.replace(rawPermalink, changedPermalinks[rawPermalink]);
+      hasFileChanged = true;
+    }
+  });
+
+  fileContent = hasFileChanged
+    ? dom.window.document.body.innerHTML
+    : fileContent;
+  const markdownRelativeUrlMatches = fileContent.match(markdownRegex);
+
+  if (markdownRelativeUrlMatches) {
+    for (const match of markdownRelativeUrlMatches) {
+      let originalPermalink = match.slice(match.indexOf("(") + 1, -1);
```

Comment:
should we do a `map` here?

Reply:
there is a break in the control flow cuz got `continue` keyword ya

Reply:
yea we could do a `filter -> map` but it's a style issue. i think this is ok for now

Reply:
72769cc true, am using return for now!